custom-elements/form-associated/ElementInternals-validation.html

```diff
@@ -0,0 +1,225 @@
+  assert_false(control.i.checkValidity());
+  assert_equals(invalidCount, 1);
+}, 'checkValidity()');
+
+test(() => {
+  const control = document.createElement('my-control');
+  document.body.appendChild(control);
+  control.tabIndex = 0;
+  let invalidCount = 0;
+  control.addEventListener('invalid', e => {
+    assert_equals(e.target, control);
+    assert_true(e.cancelable);
+    ++invalidCount;
+  });
+
+  assert_true(control.i.reportValidity(), 'default state');
+  assert_equals(invalidCount, 0);
+
+  assert_not_equals(document.activeElement, control);
+  control.i.setValidity({customError:true}, 'foo');
+  assert_false(control.i.reportValidity());
+  assert_equals(invalidCount, 1);
+  assert_equals(document.activeElement, control);
+
+  control.blur();
+  control.addEventListener('invalid', e => e.preventDefault());
+  assert_false(control.i.reportValidity());
+  assert_not_equals(document.activeElement, control);
+}, 'reportValidity()');
+
+test(() => {
+  const container = document.getElementById('container');
+  container.innerHTML = '<form><input type=submit><my-control>';
+  const form = container.querySelector('form');
+  const control = container.querySelector('my-control');
+  control.tabIndex = 0;
+
+  assert_true(control.i.checkValidity());
+  assert_true(form.checkValidity());
+  control.i.setValidity({valueMissing: true}, 'Please fill out this field');
+  assert_false(form.checkValidity());
+
+  assert_false(form.reportValidity());
+  assert_equals(document.activeElement, control);
+
+  control.blur();
+  container.querySelector('input[type=submit]').click();
+  assert_equals(document.activeElement, control);
+}, 'Custom control affects validation at the owner form');
+
+function isValid(element, comment) {
+  assert_true(element.matches(':valid'), comment ? (comment + ' - :valid') : undefined);
+  assert_false(element.matches(':invalid'), comment ? (comment + ' - :invalid') : undefined);
+}
+
+function isInvalid(element, comment) {
+  assert_false(element.matches(':valid'), comment ? (comment + ' - :valid') : undefined);
+  assert_true(element.matches(':invalid'), comment ? (comment + ' - :invalid') : undefined);
+}
+
+test(() => {
+  const container = document.getElementById('container');
+  container.innerHTML = '<form><fieldset><my-control>';
+  const form = container.querySelector('form');
+  const fieldset = container.querySelector('fieldset');
+  const control = container.querySelector('my-control');
+
+  isValid(control);
+  isValid(form);
+  isValid(fieldset);
+
+  control.i.setValidity({typeMismatch: true}, 'Invalid format');
+  isInvalid(control);
+  isInvalid(form);
+  isInvalid(fieldset);
+
+  control.remove();
+  isValid(form);
+  isValid(fieldset);
+
+  fieldset.appendChild(control);
+  isInvalid(form);
+  isInvalid(fieldset);
+
+  control.i.setValidity({});
+  isValid(control);
+  isValid(form);
+  isValid(fieldset);
+}, ':valid :invalid for FORM and FIELDSET');
+</script>
+</body>
```

Comment:
We should test the shadow-including descendant requirement on the anchor element